bmc-log-collector/log-collector.go

```diff
@@ -0,0 +1,160 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log/slog"
+	"net/http"
+	"os"
+	"path"
+	"strconv"
+	"sync"
+	"time"
+)
+
+type SystemEventLog struct {
+	Od_Id             string   `json:"@odata.id"`
+	Od_Type           string   `json:"@odata.type"`
+	Create            string   `json:"Created"`
+	Description       string   `json:"Description"`
+	EntryCode         string   `json:"EntryCode"`
+	EntryType         string   `json:"EntryType"`
+	GeneratorId       string   `json:"GeneratorId"`
+	Id                string   `json:"Id"`
+	Message           string   `json:"Message"`
+	MessageArgs       []string `json:"MessageArgs"`
+	OdCnt_MessageArgs int      `json:"MessageArgs@odata.count"`
+	MessageId         string   `json:"MessageId"`
+	Name              string   `json:"Name"`
+	SensorNumber      int      `json:"SensorNumber"`
+	SensorType        string   `json:"SensorType"`
+	Severity          string   `json:"Severity"`
+	Serial            string
+	NodeIP            string
+	BmcIP             string
+}
+
+type Redfish struct {
+	Name        string           `json:"Name"`
+	Count       int              `json:"Members@odata.count"`
+	Context     string           `json:"@odata.context"`
+	Id          string           `json:"@odata.id"`
+	Type        string           `json:"@odata.type"`
+	Description string           `json:"Descriptionta"`
+	Sel         []SystemEventLog `json:"Members"`
+}
+
+type logCollector struct {
+	machinesPath string       // Machine list path
+	rfUrl        string       // Redfish API address
+	ptrDir       string       // Pointer store
+	testMode     bool         // when testMode is true, write text file for test
+	testOut      string       // Test output directory
+	user         string       // iDRAC user
+	password     string       // iDRAC password
+	rfClient     *http.Client // to reuse HTTP transport
+	mutex        *sync.Mutex  // execlusive lock for pointer
+}
+
+func (c *logCollector) logCollectorWorker(ctx context.Context, wg *sync.WaitGroup, m Machine) {
+	defer wg.Done()
+
+	rfc := RedfishClient{
+		user:     c.user,
+		password: c.password,
+		client:   c.rfClient,
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+
+		default:
+			url := "https://" + m.BmcIP + c.rfUrl
+			byteBuf, err := rfc.requestToBmc(ctx, url)
+			if err != nil {
+				// When canceled by context, the pointer files is never updated because the return is made here.
+				errMsg := fmt.Sprintf("%s: %s", err, url)
+				slog.Error(errMsg)
```

Comment:
Is this error handling sufficient?
For example, we might need retry or metrics.

Reply:
a retry is no need. there is accessing by interval time.